lib/backend/syncersv1/updateprocessors/profileprocessor.go

```diff
@@ -0,0 +1,236 @@
+	return &profileUpdateProcessor{
+		v2Kind: apiv2.KindProfile,
+	}
+}
+
+// Need to create custom logic for Profile since it breaks the values into 3 separate KV Pairs:
+// Tags, Labels, and Rules.
+type profileUpdateProcessor struct {
+	v2Kind string
+}
+
+func (pup *profileUpdateProcessor) Process(kvp *model.KVPair) ([]*model.KVPair, error) {
+	// Check the v2 resource is the correct type.
+	rk, ok := kvp.Key.(model.ResourceKey)
+	if !ok || rk.Kind != pup.v2Kind {
+		return nil, fmt.Errorf("Incorrect key type - expecting resource of kind %s", pup.v2Kind)
+	}
+
+	// Convert the v2 resource to the equivalent v1 resource type.
+	v2key, ok := kvp.Key.(model.ResourceKey)
+	if !ok {
+		return nil, errors.New("Key is not a valid V2 resource key")
+	}
+	v1keys, err := convertProfileV2ToV1Keys(v2key)
+	if err != nil {
+		return nil, err
+	}
+
+	responses := make([]*model.KVPair, 3)
+	for i, v1key := range v1keys {
+		responses[i] = &model.KVPair{
+			Key: v1key,
+		}
+	}
+
+	var v1values []interface{}
+	// Deletion events will have a value of nil. Do not convert anything for a deletion event.
+	if kvp.Value != nil {
+		v1Profile, err := convertProfileV2ToV1Value(kvp.Value)
+		if err == nil {
+			// Currently ignore the error so that incorrect values get skipped instead of erroring out
+			v1values = append(v1values, v1Profile.Tags)
+			v1values = append(v1values, v1Profile.Labels)
+			v1values = append(v1values, &v1Profile.Rules)
+		}
+	}
+
+	// Create the KV Pairs by matching the keys with the correct values.
+	for i, v1value := range v1values {
+		responses[i].Value = v1value
+		responses[i].Revision = kvp.Revision
+	}
+
+	return responses, nil
+}
+
+func (pup *profileUpdateProcessor) OnSyncerStarting() {
+	// Do nothing
+}
+
+func convertProfileV2ToV1Keys(v2key model.ResourceKey) ([]model.Key, error) {
+	// Returns the keys in order: ProfileTagsKey, ProfileLabelsKey, ProfileRulesKey
+	if v2key.Name == "" {
+		return []model.Key{}, errors.New("Missing Name field to create a v1 Profile Key")
+	}
+	pk := model.ProfileKey{
+		Name: v2key.Name,
+	}
+	return []model.Key{
+		model.ProfileTagsKey{pk},
+		model.ProfileLabelsKey{pk},
+		model.ProfileRulesKey{pk},
+	}, nil
+
+}
+
+func convertProfileV2ToV1Value(val interface{}) (*model.Profile, error) {
+	v2res, ok := val.(*apiv2.Profile)
+	if !ok {
+		return nil, errors.New("Value is not a valid Profile resource value")
+	}
+
+	var irules []model.Rule
+	for _, irule := range v2res.Spec.IngressRules {
+		irules = append(irules, RuleAPIV2ToBackend(irule))
+	}
+
+	var erules []model.Rule
+	for _, erule := range v2res.Spec.EgressRules {
+		erules = append(erules, RuleAPIV2ToBackend(erule))
+	}
+
+	rules := model.ProfileRules{
+		InboundRules:  irules,
+		OutboundRules: erules,
+	}
+
+	v1value := &model.Profile{
+		Rules:  rules,
+		Labels: v2res.Spec.LabelsToApply,
+	}
+
+	return v1value, nil
+}
+
+func RulesAPIV2ToBackend(ars []apiv2.Rule) []model.Rule {
```

Comment:
Might be nice to split these out into a rules.go or similar since they are shared by profiles and the different network policy processors.

Reply:
not too fussed though.